drop-transfer/src/ws/server/v2.rs

```diff
@@ -465,6 +466,174 @@ impl handler::Downloader for Downloader {
 
 impl handler::Request for (v2::TransferRequest, IpAddr, Arc<DropConfig>) {
     fn parse(self) -> anyhow::Result<IncomingTransfer> {
-        self.try_into().context("Failed to parse transfer request")
+        let (v2::TransferRequest { files }, peer, config) = self;
+        Ok(IncomingTransfer::new(peer, map_files(files)?, &config)?)
+    }
+}
+
+fn map_files(files: Vec<v2::File>) -> anyhow::Result<Vec<FileToRecv>> {
+    fn process_file(
+        files: &mut Vec<(FileSubPath, u64)>,
+        subpath: FileSubPath,
+        v2::File { size, children, .. }: v2::File,
+    ) -> crate::Result<()> {
+        match size {
+            Some(size) => {
+                files.push((subpath, size));
+            }
+            None => {
+                for file in children {
+                    process_file(files, subpath.clone().append_file_name(&file.name)?, file)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    let mut info = Vec::new();
+    for file in files {
+        process_file(&mut info, FileSubPath::from_file_name(&file.name)?, file)?;
+    }
+
+    let mut files = Vec::with_capacity(info.len());
+    let mut used_mappings = HashMap::new();
+
+    for (mut path, size) in info {
+        let uroot = path.root();
+        let nroot = utils::normalize_filename(uroot);
+
+        'vars: for nvariant in utils::filepath_variants(nroot.as_ref())?
```

Comment:
`'vars`? Could you explain the syntax, never seen before in Rust

Reply:
https://doc.rust-lang.org/rust-by-example/flow_control/loop/nested.html